source/server/hot_restarting_base.cc

```diff
@@ -0,0 +1,198 @@
+    if (replyIsExpectedType(&proto, HotRestartMessage::Reply::kPassListenSocket) &&
+        proto.reply().pass_listen_socket().fd() != -1) {
+      memset(control_buffer, 0, CMSG_SPACE(sizeof(int)));
+      message.msg_control = control_buffer;
+      message.msg_controllen = CMSG_SPACE(sizeof(int));
+      cmsghdr* control_message = CMSG_FIRSTHDR(&message);
+      control_message->cmsg_level = SOL_SOCKET;
+      control_message->cmsg_type = SCM_RIGHTS;
+      control_message->cmsg_len = CMSG_LEN(sizeof(int));
+      *reinterpret_cast<int*>(CMSG_DATA(control_message)) = proto.reply().pass_listen_socket().fd();
+      ASSERT(sent == total_size, "an fd passing message was too long for one sendmsg().");
+    }
+
+    int rc = sendmsg(my_domain_socket_, &message, 0);
+    RELEASE_ASSERT(rc == static_cast<int>(cur_chunk_size),
+                   fmt::format("hot restart sendmsg() failed: returned {}, errno {}", rc, errno));
+  }
+  RELEASE_ASSERT(fcntl(my_domain_socket_, F_SETFL, O_NONBLOCK) != -1,
+                 fmt::format("Set domain socket nonblocking failed, errno = {}", errno));
+}
+
+bool HotRestartingBase::replyIsExpectedType(const HotRestartMessage* proto,
+                                            HotRestartMessage::Reply::ReplyCase oneof_type) {
+  return proto != nullptr && proto->requestreply_case() == HotRestartMessage::kReply &&
+         proto->reply().reply_case() == oneof_type;
+}
+
+// Pull the cloned fd, if present, out of the control data and write it into the
+// PassListenSocketReply proto; the higher level code will see a listening fd that Just Works. We
+// should only get control data in a PassListenSocketReply, it should only be the fd passing type,
+// and there should only be one at a time. Crash on any other control data.
+void HotRestartingBase::getPassedFdIfPresent(HotRestartMessage* out, msghdr* message) {
+  cmsghdr* cmsg = CMSG_FIRSTHDR(message);
+  if (cmsg != nullptr) {
+    RELEASE_ASSERT(cmsg->cmsg_level == SOL_SOCKET && cmsg->cmsg_type == SCM_RIGHTS &&
+                       replyIsExpectedType(out, HotRestartMessage::Reply::kPassListenSocket),
+                   "recvmsg() came with control data when the message's purpose was not to pass a "
+                   "file descriptor.");
+
+    out->mutable_reply()->mutable_pass_listen_socket()->set_fd(
+        *reinterpret_cast<int*>(CMSG_DATA(cmsg)));
+
+    RELEASE_ASSERT(CMSG_NXTHDR(message, cmsg) == nullptr,
+                   "More than one control data on a single hot restart recvmsg().");
+  }
+}
+
+std::unique_ptr<HotRestartMessage> HotRestartingBase::receiveHotRestartMessage(Blocking block) {
+  // By default the domain socket is non blocking. If we need to block, make it blocking first.
+  if (block == Blocking::Yes) {
+    RELEASE_ASSERT(fcntl(my_domain_socket_, F_SETFL, 0) != -1,
+                   fmt::format("Set domain socket blocking failed, errno = {}", errno));
+  }
+  if (recv_buf_.size() < MaxSendmsgSize) {
```

Comment:
Something about this seems not right. Would it be more accurate to say `recv_buf_.empty()` here as the if condition? Maybe also add a comment?

Reply:
Hmm. Yeah, the (unnecessary) `cur_msg_recvd_bytes_ = 0` part makes this not just a defensive "make sure its big enough", but rather real logic. I split a couple of things out into helper functions to make it clearer.